analyzer/src/main/kotlin/managers/Npm.kt

```diff
@@ -389,67 +371,87 @@ open class Npm(
             )
         }
 
+    private fun parseDependencies(moduleDir: File, scopes: Set<String>): SortedSet<PackageReference> =
+        buildTree(moduleDir, scopes)!!.dependencies
+
     private fun buildTree(
-        rootModulesDir: File, startModulesDir: File, moduleName: String, packages: Map<String, Package>,
-        dependencyBranch: List<String> = emptyList()
+        moduleDir: File,
+        scopes: Set<String>,
+        ancestorModuleDirs: List<File> = emptyList(),
+        ancestorModuleIds: Set<Identifier> = emptySet(),
+        packageType: String = managerName
     ): PackageReference? {
-        log.debug { "Building dependency tree for '$moduleName' from directory '$startModulesDir'." }
-
-        val packageFile = startModulesDir.resolve(moduleName).resolve("package.json")
+        val moduleInfo = getModuleInfo(moduleDir, scopes)
+        val dependencies = sortedSetOf<PackageReference>()
+        val moduleId = splitNamespaceAndName(moduleInfo.name).let { (namespace, name) ->
+            Identifier(packageType, namespace, name, moduleInfo.version)
+        }
 
-        if (packageFile.isFile) {
-            log.debug { "Found package file for module '$moduleName' at '$packageFile'." }
+        if (ancestorModuleIds.contains(moduleId)) {
+            val cycle = ancestorModuleIds.toList().let {
+                val cycleStartIndex = it.indexOf(moduleId)
+                it.subList(cycleStartIndex, it.size) + moduleId
+            }.joinToString(" -> ")
```

Comment:
Given that `ancestorModuleIds` is a set, is it guaranteed that this lists the modules in the right order? Would it be better to make it a list instead?

Reply:
As a `LinkedHashSet` is use I think the order should be correct.

Reply:
Yes, either change the type to `LinkedHashSet` or `List`, because order matters and that is important to the reader.

Reply:
Done.